### PR TITLE
Add data collection configuration for Kubernetes

### PR DESCRIPTION
**Pull request by flands**
Also add language around Splunk Observability versus Platform

### CODE REVIEW

File: specification/configuration.md [hunk 1]
```diff
@@ -89,6 +90,50 @@ environment variables:
...
+environment variables, package management solutions such as Helm charts and
+Operators require YAML-based configuration. As a result, Kubernetes package
+management solutions MUST support the following YAML configuration options:
+
+- `splunkCommon`
+  - `clusterName` ()          : [REQUIRED] Name of the cluster.
+  - `distro` ()               : Which distribution of Kubernetes is deployed.
+  - `environment` ()          : Name of the environment; if not defined then skipped.
+  - `logsEngine` (`fluentd`)  : How to collect logs.
+  - `modes`
```
Comment (dmitryax) [inline specification/configuration.md:105]: I don't agree that this is a `mode`. Helm can install the connector in many modes, so it's rather a component that can be enabled or disabled than a mode as in other single-component packaging solutions.

File: specification/configuration.md [hunk 2]
```diff
@@ -89,6 +90,50 @@ environment variables:
...
   defined.
 
+### Kubernetes
+
+While Kubernetes supports container technology that can be configured using
+environment variables, package management solutions such as Helm charts and
+Operators require YAML-based configuration. As a result, Kubernetes package
+management solutions MUST support the following YAML configuration options:
+
+- `splunkCommon`
```
Comment (dmitryax) [inline specification/configuration.md:100]: Assuming we target helm initially in this specification. The existing configuration interface in the helm chart tend to be flat and don't use grouping unless it's necessary following helm best practices https://helm.sh/docs/chart_best_practices/values/#flat-or-nested-values

Do we need this particular grouping under `splunkCommon`? Can we assume any flat configuration other than `splunk<Observability|Platform>` considered to be common?

File: specification/configuration.md [hunk 3]
```diff
@@ -89,6 +90,50 @@ environment variables:
...
+In addition, the following Kubernetes secret configuration options MUST be
+supported:
+
+- Splunk Observability
+  - `splunk_observability_access_token`
+- Splunk Platform
+  - `splunk_platform_hec_token`
+  - `hec_client_cert`
+  - `hec_client_key`
+  - `hec_ca_file`
```
Comment (dmitryax) [inline specification/configuration.md:135]: I think we need to add `splunk_platform_` prefix to all of these. Splunk Observability doesn't support ,them but they would be different anyway

File: specification/configuration.md [hunk 4]
```diff
@@ -89,6 +90,50 @@ environment variables:
...
+### Kubernetes
+
+While Kubernetes supports container technology that can be configured using
+environment variables, package management solutions such as Helm charts and
+Operators require YAML-based configuration. As a result, Kubernetes package
+management solutions MUST support the following YAML configuration options:
+
+- `splunkCommon`
+  - `clusterName` ()          : [REQUIRED] Name of the cluster.
+  - `distro` ()               : Which distribution of Kubernetes is deployed.
```
Comment (dmitryax) [inline specification/configuration.md:102]: `platform` is needed here as well I believe

File: specification/configuration.md [hunk 5]
```diff
@@ -89,6 +90,50 @@ environment variables:
...
+  - `environment` ()          : Name of the environment; if not defined then skipped.
+  - `logsEngine` (`fluentd`)  : How to collect logs.
+  - `modes`
+    - `agent`                 : Deployed as a DaemonSet.
+      - `config` ()           : Updates configuration. Non-list options merged, list options override.
+    - `k8sClusterReceiver`    : Deployed as a standalone service.
+      - `config` ()           : Updates configuration. Non-list options merged, list options override.
```
Comment (dmitryax) [inline specification/configuration.md:109]: Both of them need `enabled` option as well. For example, they both can be disabled in order to deploy gateway-only mode

File: specification/configuration.md (hunk 1 above)
Comment (flands) [reply specification/configuration.md:105]: In our docs today, we call it a mode: https://github.com/signalfx/splunk-otel-collector/blob/main/docs/agent-vs-gateway.md. I think we should be consistent. How about `deployment`?

File: specification/configuration.md (hunk 2 above)
Comment (flands) [reply specification/configuration.md:100]: The guidance in the link appears to go both way. For example: `When there are a large number of related variables, and at least one of them is non-optional, nested values may be used to improve readability.` My concern is that there are a lot of additional options and lack of nesting for confusing for me.

What if we do:

```
splunk:
  environment:
  observability:
    accessToken:
  platform:
    token:
```

File: specification/configuration.md (hunk 3 above)
Comment (flands) [reply specification/configuration.md:135]: well LO supports hec... maybe just prefix with `splunk_`?

File: specification/configuration.md (hunk 4 above)
Comment (flands) [reply specification/configuration.md:102]: We need platform, provider, and distro?? Are all required? Do we have a default for any?

File: specification/configuration.md (hunk 5 above)
Comment (flands) [reply specification/configuration.md:109]: Ah, makes sense.

File: specification/configuration.md (hunk 1 above)
Comment (dmitryax) [reply specification/configuration.md:105]: Can we keep "mode" term for single-component deployments and use something else like "component" multi-component deployments?

"deployment" doesn't seem to be a good choice to me because it can be confused with k8s deployment resource, and "agent" is a daemonset not deployment.

Looking at the customer experience so far the most common usage is adjusting recourses and adding extra components to the default agent's config. With the current helm chart, their custom helm values file would look like this:

```yaml
splunkObservability:
  accessToken: xxx
  realm: us0
otelAgent:
  resources:
    limits:
      cpu: 500m
      memory: 1Gi
  config:
    receivers:
      smartagent/docker-container-stats:
        type: docker-container-stats
        dockerURL: unix:///hostfs/var/run/docker.sock
    service:
      pipelines:
        metrics:
          receivers:
            - smartagent/docker-container-stats
            - hostmetrics
            - kubeletstats
            - otlp
            - receiver_creator
            - signalfx
```

It's already pretty deep nesting for agent configuration changes, and I heard about many cases when users confuse the layers. If we add 2 more levels on top of it (`splunkCommon` -> `modes`), I believe it'll make user experience worse.

And many of configurations options specific to helm chart release and kubernetes resources are on the top level because they already have established naming patterns in helm ecosystem, examples:

```
nameOverride: ""
fullnameOverride: ""
serviceAccount:
secret:
rbac:
```

Those options are pretty well known to those who used other helm chart before. So I think we should keep them at the top level. 

To summarize, I would really recommend to keep the components groups at the top level as well. May be we can prefix them with some common word, something like `componentAgent`, `componentGateway` and `componentK8sClusterReceiver`? I took `otel` prefix because we still have `fluent` as another component, so to keep the distinction now we have `otelAgent`, `otelCollector`, `otelK8sClusterReceiver` and `fluentd`. And for me it looks better than other alternatives, assuming `otelCollector` is renamed to `otelGateway`

File: specification/configuration.md (hunk 3 above)
Comment (dmitryax) [reply specification/configuration.md:135]: LO supports hec, but can we assume that the same certificates always must be used for both backends?

File: specification/configuration.md (hunk 4 above)
Comment (dmitryax) [reply specification/configuration.md:102]: Oh sorry, I mean `provider`, not `platform`. We need `provider` and `distro`, both of them are optional

File: specification/configuration.md (hunk 2 above)
Comment (owais) [reply specification/configuration.md:100]: +1 this is what the operator does as well i.e, it accepts "common" fields at the root level and then has nested levels for different components e.g, 

```yaml
spec:
  clusterName: <MY_CLUSTER_NAME>
  realm: <SPLUNK_REALM>
  agent:
     ...
  gateway:
     ...
```

https://github.com/signalfx/splunk-otel-collector-operator/blob/main/docs/splunkotelagent_cr_spec.md

File: specification/configuration.md (hunk 1 above)
Comment (owais) [reply specification/configuration.md:105]: FWIW "mode" does sound a bit confusing to me as well. It sounds as if I can only pick one of the three modes especially if I'm using a single/common config file to deploy multiple agents.

File: specification/configuration.md [hunk 6]
```diff
@@ -89,6 +90,53 @@ environment variables:
...
+### Kubernetes
+
+While Kubernetes supports container technology that can be configured using
+environment variables, package management solutions such as Helm charts and
+Operators require YAML-based configuration. As a result, Kubernetes package
+management solutions MUST support the following YAML configuration options:
+
+- `clusterName` ()            : [REQUIRED] Name of the cluster.
+- `environment` ()            : Name of the environment; if not defined then skipped.
+- `distro` ()                 : Which distribution of Kubernetes is deployed.
```
Comment (dmitryax) [inline specification/configuration.md:102]: Maybe full name `distribution`?

File: specification/configuration.md [hunk 7]
```diff
@@ -89,6 +90,53 @@ environment variables:
...
+- `provider` ()               : Where Kubernetes is deployed.
```
Comment (dmitryax) [inline specification/configuration.md:103]: Don't have strong preference but maybe we should consider `cloudProvider`?

**Review by dmitryax — APPROVED** [review]
LGTM

File: specification/configuration.md [hunk 8]
```diff
@@ -89,6 +90,52 @@ environment variables:
...
   defined.
 
+### Kubernetes
+
+While Kubernetes supports container technology that can be configured using
+environment variables, package management solutions such as Helm charts and
+Operators require YAML-based configuration. As a result, Kubernetes package
+management solutions MUST support the following YAML configuration options:
+
+- `clusterName` ()             : [REQUIRED] Name of the cluster.
```
Comment (dmitryax) [inline specification/configuration.md:100]: This is actually required only if Observability destination is enabled. I think we should mention that

File: specification/configuration.md [hunk 9]
```diff
@@ -89,6 +90,52 @@ environment variables:
   If `SPLUNK_CONFIG` is defined then these options are not required but MAY be
   defined.
 
+### Kubernetes
```
Comment (MrAlias) [inline specification/configuration.md:93]:
```suggestion
### Kubernetes Package Management Solutions
```

File: specification/configuration.md [hunk 10]
```diff
@@ -89,6 +90,52 @@ environment variables:
...
+  - `enabled` (`true`)         : Whether agent is deployed.
+  - `config` ()                : Updates configuration. Non-list options merged, list options override.
+- `gateway`                    : Deployed as a clustered Service and receives data from agent.
+  - `enabled` (`false`)        : Whether gateway is deployed.
+  - `config` ()                : Updates configuration. Non-list options merged, list options override.
+- `clusterReceiver`            : Deployed as a single replica deployment and collects Kubernetes API cluster and event telemetry.
+  - `enabled` (`true`)         : Whether k8sClusteReceiver is deployed. Ignored if `metricsEnabled` is `false`.
+  - `config` ()                : Updates configuration. Non-list options merged, list options override.
+
+In addition, at least one of the following sections MUST be specified:
```
Comment (MrAlias) [inline specification/configuration.md:114]: "following sections" is a bit vague, it can be interpreted as a paragraph or markdown section as well as a level in the following list. Can we nail this down?

```suggestion
In addition, at least one of the the following configuration groups, `splunkObservability` or `splunkPlatform`, MUST be specified:
```

File: specification/configuration.md [hunk 11]
```diff
@@ -89,6 +90,52 @@ environment variables:
...
+Finally, the following Kubernetes secret configuration options MUST be
+supported:
+
+- Splunk Observability
+  - `splunk_observability_access_token`
+- Splunk Platform
+  - `splunk_platform_hec_token`
+  - `splunk_platform_hec_client_cert`
+  - `splunk_platform_hec_client_key`
+  - `splunk_platform_hec_ca_file`
```
Comment (MrAlias) [inline specification/configuration.md:137]: Should these be interwoven with the above list? They seem subsets of the above configuration options, and potentially duplicates (i.e. `splunk_observability_access_token` looks very similar to `accessToken` in the `splunkObservability` list element)

Also, should they include descriptions and requirement levels?

File: specification/configuration.md (hunk 8 above)
Comment (MrAlias) [reply specification/configuration.md:100]: Does it make sense to the authors of Kuberntetes package management solutions what required means? Do we need to define behavior for them when a required variable is not provided?

File: specification/configuration.md (hunk 11 above)
Comment (flands) [reply specification/configuration.md:137]: This is specific for k8s secrets. @dmitryax if you specify this do you need to specify say `accessToken` above?

File: specification/configuration.md [hunk 12]
```diff
@@ -89,6 +90,52 @@ environment variables:
   If `SPLUNK_CONFIG` is defined then these options are not required but MAY be
   defined.
 
+### Kubernetes Package Management Solutions
+
+While Kubernetes supports container technology that can be configured using
+environment variables, package management solutions such as Helm charts and
+Operators require YAML-based configuration. As a result, Kubernetes package
+management solutions MUST support the following YAML configuration options:
```
Comment (flands) [inline specification/configuration.md:98]:
```suggestion
management solutions MUST support the YAML configuration options specified below.

> Any option description listed as REQUIRED means a value for the option MUST be specified.
```

File: specification/configuration.md [hunk 13]
```diff
@@ -89,6 +90,52 @@ environment variables:
...
+  - `enabled` (`true`)         : Whether agent is deployed.
+  - `config` ()                : Updates configuration. Non-list options merged, list options override.
+- `gateway`                    : Deployed as a clustered Service and receives data from agent.
+  - `enabled` (`false`)        : Whether gateway is deployed.
+  - `config` ()                : Updates configuration. Non-list options merged, list options override.
+- `clusterReceiver`            : Deployed as a single replica deployment and collects Kubernetes API cluster and event telemetry.
+  - `enabled` (`true`)         : Whether k8sClusteReceiver is deployed. Ignored if `metricsEnabled` is `false`.
+  - `config` ()                : Updates configuration. Non-list options merged, list options override.
+
+In addition, at least one of the the following configuration groups, `splunkObservability` or `splunkPlatform`, MUST be specified:
```
Comment (flands) [inline specification/configuration.md:114]:
```suggestion
In addition, at least one of the below configuration groups, 
`splunkObservability` or `splunkPlatform`, MUST be specified.

> Any option description listed as REQUIRED means a value for the option MUST be specified.
```

File: specification/configuration.md (hunk 8 above)
Comment (flands) [reply specification/configuration.md:102]: Addressed your comment @MrAlias . @dmitryax I think we should be consistent -- either both require or both don't.

File: specification/configuration.md (hunk 11 above)
Comment (dmitryax) [reply specification/configuration.md:142]: > This is specific for k8s secrets. @dmitryax if you specify this do you need to specify say accessToken above?
No, the secrets by default are set from yaml configuration, e.g. `splunkObservability.accessToken` goes to `splunk_observability_access_token`. But, if user wants to specify secrets by themselves, they can set them directly (`splunk_observability_access_token`) -- in that case `splunkObservability.accessToken` can be omitted.

File: specification/configuration.md (hunk 8 above)
Comment (dmitryax) [reply specification/configuration.md:102]: Ok, I'll check on this with SCK team

File: specification/configuration.md (hunk 11 above)
Comment (MrAlias) [reply specification/configuration.md:142]: > > This is specific for k8s secrets. @dmitryax if you specify this do you need to specify say accessToken above?
> 
> No, the secrets by default are set from yaml configuration, e.g. `splunkObservability.accessToken` goes to `splunk_observability_access_token`. But, if user wants to specify secrets by themselves, they can set them directly (`splunk_observability_access_token`) -- in that case `splunkObservability.accessToken` can be omitted.

Do we need an additional `Kubernetes` section? It seems like this section might be serving two topics? Also, would we not prefer environment variables for native kube configuration of the Splunk observability access token?

Comment (flands) [reply specification/configuration.md]: I updated the previous section to hopefully clarify.

File: specification/configuration.md [hunk 14]
```diff
@@ -89,6 +90,64 @@ environment variables:
...
+  - `config` ()                : Updates configuration. Non-list options merged, list options override.
+- `clusterReceiver`            : Deployed as a single replica deployment and collects Kubernetes API cluster and event telemetry.
+  - `enabled` (`true`)         : Whether k8sClusteReceiver is deployed. Ignored if `metricsEnabled` is `false`.
+  - `config` ()                : Updates configuration. Non-list options merged, list options override.
+
+In addition, at least one of the below configuration groups,
+`splunkObservability` or `splunkPlatform`, MUST be specified.
+
+> Any option description listed as REQUIRED means a value for the option MUST
+> be specified. Both `accessToken` and `token` values are stored as Kubernetes
```
Comment (dmitryax) [inline specification/configuration.md:122]: Should we clarify that `[REQUIRED]` here means that the option is required only within a particular group?

**Review by dmitryax — APPROVED** [review]

File: specification/configuration.md [hunk 15]
```diff
@@ -89,6 +90,64 @@ environment variables:
...
+`splunkObservability` or `splunkPlatform`, MUST be specified.
+
+> Any option description listed as REQUIRED means a value for the option MUST
+> be specified in that section. Both `accessToken` and `token` values are
+> stored as Kubernetes secrets. The secret key names are specified after this
+> section. If Kubernetes secret keys are specified they will override the
+> values specified below. In short, `accessToken` and `token` are required only
+> if not specified in Kubernetes secrets.
+
+- `splunkObservability`
```
Comment (owais) [inline specification/configuration.md:128]: nit: generally config like this should be namespaced/encapsulated so using `splunk` prefix is probably redundant. Can we call these just `observability` and `platform/enterprise`?

**Review by owais — COMMENTED** [review]
Looks good. Left a nit-picky suggestion.

Q: Do we want to specify k8s API namespace(s) here? These are used whenever we install any custom resource definition in a k8s cluster. For example, the operator today uses `o11y.splunk.com` as the API namespace for Otel deployments. Do we want to prescribe this or out of scope for this PR? ref: https://github.com/signalfx/splunk-otel-collector-operator/blob/main/docs/splunkotelagent_cr_spec.md#splunkotelagent-custom-resource-specification

File: specification/configuration.md (hunk 15 above)
Comment (MrAlias) [reply specification/configuration.md:128]: It's been a while since I used helm, so I might be missing something, but as I remember it the namespace is defined by the operator when they install the chart. Meaning it is up to them to correctly define it as `splunk`, right? If so, it might make sense to keep the prefix, even though it is redundant, to ensure a clear scope to the variable.

**Review by MrAlias — APPROVED** [review]

File: specification/configuration.md (hunk 15 above)
Comment (dmitryax) [reply specification/configuration.md:128]: It's not really a prefix for Splunk specific configuration, but rather part of a destination configuration. Full names for the destinations/products are "Splunk Observability" and "Splunk Platform" (https://www.splunk.com/en_us/software.html). `observability` may be self-descriptive, but just `platform` is confusing I think. I would keep the full names to make it clear.

Comment (owais) [reply specification/configuration.md:128]: That sounds reasonable and not a blocker for me but to answer the question,

> It's been a while since I used helm, so I might be missing something, but as I remember it the namespace is defined by the operator when they install the chart. Meaning it is up to them to correctly define it as splunk, right? 

My bad for using the term "namespace". I did not mean a k8s namespace but namespace in a general sense.

I was thinking of somethig like this:

```
k get splunk-agent

kind: SplunkOtelAgent
spec:
    clusterName: ...
    splunkObservability: ...
    splunkPlatform: ...
```

here repeating the word splunk in config feels unnecessary. 

Is this not the case for helm? Does splunk config get merged into an object that contains non-splunk config?

Comment (dmitryax) [reply specification/configuration.md:128]: The configuration of the helm is applied to the whole chart release, it's on the root level, not a part of any object. There are some Splunk specific configs and Kubernetes specific configs in it. We already discussed extra grouping in this thread https://github.com/signalfx/gdi-specification/pull/147#discussion_r742334481. And I think we agreed that extra grouping like `splunkCommon` will make the user experience worse.

Looking at the same configuration in the `SplunkOtelAgent` k8s object, I think it's Ok to have some extra Splunk words in it given that `splunkPlatform` and `splunkObservability` refer to exact product names that IMO shouldn't be broken down.

**Review by owais — APPROVED** [review]
